README.md
# Curbside Pickup plugin for Koha

## Downloading
From the [release page](https://github.com/bywatersolutions/koha-plugin-curbside-pickup/releases) you can download the last KPZ file.

## Configuration
The plugin configuration has a tab for each library set up in your Koha instance.
Each tab the following options:

### Enable
Simply put, the option enables curbside pickups for this libxrary.

### Pickup interval
This is the amount of time allotted for each pickup.
For example, if the value is set to `30`, each pickup slot will be 30 minutes long.
For best results, use whole fractions of an hour ( 10, 15, 20, 30, 60 ).

### Maximum patrons per interval
This is the number of simultaneous pickups that may occupy a single time slot.
For example, if it is set to 3, then up to three patrons may have pickups scheduled for the same pickup time.

### Patron-scheduled pickup
If this option is enabled, patrons may schedule their own pickup times.
If not enabled, librarians can still schedule pickups for patrons from the staff interface.

### Curbside pickup hours
This section lets you specify the start and end time for curbside pickups for each day of the week.
The time is specificed in 24 hour format ( e.g. 0:00 to 23:59 ).
Leave all four boxes empty for days that you do not want to have pickup hours.

## Usage
To use the plugin, browse to the plugins page and select the `Tool` option for the plugin.
For your convenience, this plugin also adds a 'Schedule pickup' button to the patrons toolbar that will allow you to schedule a pickup for that patron without searching for them first.

### Workflow
Each pickup moves through a lifecycle as follows:

#### To be staged
This is a brand new pickup request.
Any pickup in this status needs waiting holds for the pickup to be gathered.
Once materials are gathered, click `Mark as staged & ready` to move to the next step.

#### Staged & ready
This status indicates the materials have been gathered and are waiting for the patron.
At this stage, either the patron or a librarian can mark the pickup as "Patron has arrived" to indicate they are ready and waiting.

#### Patron is outside
At this status, it is time to take the items out to the patron.
The `Mark as delivered` button will check out the items to the patron and mark the pickup as delivered.

#### Delivered today
This tab lists all pickups completed today.

#### Schedule pickup
This tab contains the form to schedule a new curbside pickup.
Simply search for the patron's cardnumber or use the "Schedule pickup" button on the patron record.
Choose the pickup day and a list of times will display.
To the right of the given time, in parentheses is the number of pickups already scheduled for that time slot.

When a pickup is scheduled, Koha will look for a `CURBSIDE` notice in the Holds module.
If this notice exists it will be sent to the patron.
This notice supports email and SMS.
A default version of this notice should have been installed by the plugin.

## Acknowledgments
This plugin was originally developed following a specification and design
created by staff members from the Equinox Open Library Initiative, including
Andrea Buntz Neiman, Sally Fortin, and Galen Charlton.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.29] - 2021-02-11
## Changed
- Add support for handling 'phone' transports gracefully

## [1.0.28] - 2021-01-23
## Fixed
- Fix bug causing crash if a curbside pickup checkout is returned the same day.

## [1.0.27] - 2021-01-04
## Changed
- Add return to the same tab that you were on when changing a pickup status

## [1.0.26] - 2021-01-04
## Added
- Add classes to all "action" buttons for styling purposes

## [1.0.25] - 2020-12-18
## Changed
- Fixed bug related to new curbside_pickups_issues table where schema wasn't being loaded

## [1.0.24] - 2020-12-16
## Changed
- Added table curbside_pickups_issues to map a pickup to the issues & holds associated with a pickup

## [1.0.23] - 2020-12-10
## Changed
- Fix bug in OPAC that prevents patron from scheduling a new pickup unless all previous pickups are canceled/deleted

## [1.0.22] - 2020-11-20
## Changed
- Fix bug in OPAC that prevents patron from scheduling a new pickup unless all previous pickups are canceled/deleted

## [1.0.21] - 2020-11-18
## Changed
- Fix bug in OPAC javascript the prevents patron from scheduling a pickup

## [1.0.20] - 2020-11-04
## Changed
- Renamed OPAC javascript var in opac.js to avoid conflict with a var of the same name in Koha v20.05

## [1.0.17] - 2020-09-10
## Added
- Plugin now supports internationaliztion!
- Spanish translation

## [1.0.16] - 2020-09-01
## Changed
- Only show branches available for pickup in OPAC library selector.

## [1.0.15] - 2020-09-01
## Changed
- Fix bug caused by missing js variable

## [1.0.14] - 2020-08-31
## Changed
- Add 'loading' modal to OPAC while data is fetched for display

## [1.0.13] - 2020-08-27
## Changed
- Require unit tests to pass before building and releasing

## [1.0.12] - 2020-08-27
## Changed
- Update this changelog
- Remove installation step that is no longer required

## [1.0.11] - 2020-08-27
## Changed
- Add acknowledgements for original specification and design

## [1.0.10] - 2020-08-27
## Changed
- Plugin no longer requires bug 26163 to function

## [1.0.9] - 2020-08-17
## Changed
- Filter out holds from status tabs that are not waiting at the logged in library
- Disable submit button on staff side until a valid pickup slot is selected
- Disable submit button on opac side until a valid pickup slot is selected
- Ensure datepicker icon is disable correctly
- Add 'require' lines to API controller to ensure it has access to those modules

## [1.0.8] - 2020-08-17
## Added
- Specify the plugin module for the GitHub Action that builds the kpz file

## [1.0.7] - 2020-08-17
## Added
- Added documentation for CURBSIDE notice
- Added hyperlinks for various elements

## [1.0.6] - 2020-08-17
## Added
- Added default CURBSIDE notice
- Hours and minutes in configuration are now formatted correctly for 24-hour time

## [1.0.5] - 2020-08-17
## Changed
- Added support for CURBSIDE notice, which will be sent upon initial creation of a curbside pickup

## [1.0.4] - 2020-08-12
## Changed
- Remove dangling code, make fields required, prevent ISE when patron not found.

## [1.0.3] - 2020-08-11
## Added
- This changelog!

## [1.0.2] - 2020-08-11
### Changed
- Switched from `use` to `require` to allow installation to proceed without errors. Still requires plack to be restarted before using the plugin.
### Removed
- Installation debugging code.

## [1.0.1] - 2020-08-11
### Added
- Debugging for installation subroutine.

## [1.0.0] - 2020-08-11
### Added
- Initial release!



Koha/Plugin/Com/ByWaterSolutions/CurbsidePickup.pm
package Koha::Plugin::Com::ByWaterSolutions::CurbsidePickup;

use Modern::Perl;

use Mojo::JSON qw(decode_json to_json);

use base qw(Koha::Plugins::Base);

use Cwd qw(abs_path);
use Encode qw(decode);
use File::Slurp qw(read_file);
use Module::Metadata;

use C4::Auth;
use C4::Circulation;
use C4::Context;
use C4::Installer qw(TableExists);
use Koha::DateUtils;
use Koha::Libraries;
use Koha::Schema;

BEGIN {
    my $path = Module::Metadata->find_module_by_name(__PACKAGE__);
    $path =~ s!\.pm$!/lib!;
    unshift @INC, $path;

    require Koha::CurbsidePickupIssues;
    require Koha::CurbsidePickupPolicies;
    require Koha::CurbsidePickups;
    require Koha::Schema::Result::CurbsidePickup;
    require Koha::Schema::Result::CurbsidePickupIssue;
    require Koha::Schema::Result::CurbsidePickupPolicy;

    # register the additional schema classes
    Koha::Schema->register_class(CurbsidePickup => 'Koha::Schema::Result::CurbsidePickup');
    Koha::Schema->register_class(CurbsidePickupPolicy => 'Koha::Schema::Result::CurbsidePickupPolicy');
    Koha::Schema->register_class(CurbsidePickupIssue => 'Koha::Schema::Result::CurbsidePickupIssue');
    # ... and force a refresh of the database handle so that it includes
    # the new classes
    Koha::Database->schema({ new => 1 });
}

our $VERSION         = "{VERSION}";
our $MINIMUM_VERSION = "{MINIMUM_VERSION}";

our $metadata = {
    name            => 'Curbside Pickup',
    author          => 'Kyle M Hall, ByWater Solutions',
    date_authored   => '2020-08-04',
    date_updated    => "1900-01-01",
    minimum_version => $MINIMUM_VERSION,
    maximum_version => undef,
    version         => $VERSION,
    description     => 'This plugin adds curbside pickup of holds to Koha',
};

sub new {
    my ( $class, $args ) = @_;

    $args->{'metadata'} = $metadata;
    $args->{'metadata'}->{'class'} = $class;

    my $self = $class->SUPER::new($args);

    return $self;
}

sub tool {
    my ( $self, $args ) = @_;

    my $cgi      = $self->{'cgi'};
    my $template = $self->get_template( { file => 'tool.tt' } );

    my $branchcode = C4::Context->userenv->{branch};

    my $action = $cgi->param('action');

    my $CurbsidePickupPolicy = Koha::CurbsidePickupPolicies->find( { branchcode => $branchcode } );
    $template->param( policy => $CurbsidePickupPolicy );

    if ( $action eq 'find-patron' ) {
        my $cardnumber     = $cgi->param('cardnumber');
        my $borrowernumber = $cgi->param('borrowernumber');

        my $patron =
          $cardnumber
          ? Koha::Patrons->find( { cardnumber => $cardnumber } )
          : Koha::Patrons->find($borrowernumber);

        my $existing_curbside_pickups;

        if ( $patron ){
            $existing_curbside_pickups = Koha::CurbsidePickups->search(
                {
                    branchcode                => $branchcode,
                    borrowernumber            => $patron->id,
                    delivered_datetime        => undef,
                    scheduled_pickup_datetime => { '>' => \'DATE(NOW())' },
                }
            );
        } else {
            $template->param( problem => {
                type => 'no_patron_found',
                cardnumber => $cardnumber
            });
        }

        $template->param(
            patron      => $patron,
            tab         => 'schedule-pickup',
            policy_json => to_json( $CurbsidePickupPolicy->TO_JSON ),
            existing_curbside_pickups => $existing_curbside_pickups,
        );
    }
    elsif ( $action eq 'create-pickup' ) {
        my $borrowernumber            = $cgi->param('borrowernumber');
        my $scheduled_pickup_datetime = $cgi->param('pickup_time');
        my $notes                     = $cgi->param('notes');

        my $existing_curbside_pickups = Koha::CurbsidePickups->search(
            {
                branchcode                => $branchcode,
                borrowernumber            => $borrowernumber,
                delivered_datetime        => undef,
                scheduled_pickup_datetime => { '>' => \'DATE(NOW())' },
            }
        )->count();

        if ($existing_curbside_pickups) {
            $template->param(
                problem => {
                    type   => 'too_many_pickups',
                    patron => Koha::Patrons->find($borrowernumber)
                }
            );
        }
        else {
            my $curbside_pickup = Koha::CurbsidePickup->new(
                {
                    branchcode                => $branchcode,
                    borrowernumber            => $borrowernumber,
                    scheduled_pickup_datetime => $scheduled_pickup_datetime,
                    notes                     => $notes,
                }
            )->store();
            $self->_notify_new_pickup($curbside_pickup);
        }
    }
    elsif ( $action eq 'cancel' ) {
        my $id              = $cgi->param('id');
        my $curbside_pickup = Koha::CurbsidePickups->find($id);
        $curbside_pickup->delete() if $curbside_pickup;
    }
    elsif ( $action eq 'mark-as-staged' ) {
        my $id              = $cgi->param('id');
        my $curbside_pickup = Koha::CurbsidePickups->find($id);
        $curbside_pickup->set(
            {
                staged_datetime  => dt_from_string(),
                staged_by        => C4::Context->userenv->{number},
                arrival_datetime => undef,
            }
        )->store()
          if $curbside_pickup;
    }
    elsif ( $action eq 'mark-as-unstaged' ) {
        my $id              = $cgi->param('id');
        my $curbside_pickup = Koha::CurbsidePickups->find($id);
        $curbside_pickup->set(
            {
                staged_datetime  => undef,
                staged_by        => undef,
                arrival_datetime => undef,
            }
        )->store()
          if $curbside_pickup;
    }
    elsif ( $action eq 'mark-patron-has-arrived' ) {
        my $id              = $cgi->param('id');
        my $curbside_pickup = Koha::CurbsidePickups->find($id);
        $curbside_pickup->set(
            {
                arrival_datetime => dt_from_string(),
            }
        )->store()
          if $curbside_pickup;
    }
    elsif ( $action eq 'mark-as-delivered' ) {
        my $id = $cgi->param('id');

        my $curbside_pickup = Koha::CurbsidePickups->find($id);
        my $patron          = $curbside_pickup->patron;
        my $holds           = $patron->holds;
        foreach my $hold ( $holds->as_list ) {
            my @not_checked_out;
            if ( $hold->found eq 'W' && $hold->branchcode eq $branchcode ) {
                my ( $issuingimpossible, $needsconfirmation ) =
                  CanBookBeIssued( $patron, $hold->item->barcode );

                unless ( keys %$issuingimpossible ) {
                    my $issue = AddIssue( $patron->unblessed, $hold->item->barcode );
                    if ( $issue ) {
                        Koha::CurbsidePickupIssue->new({
                            curbside_pickup_id => $id,
                            issue_id           => $issue->id,
                            reserve_id         => $hold->id,
                        })->store();
                    } else {
                        push( @not_checked_out, $hold );
                    }
                }
            }

            if (@not_checked_out) {
                $template->param(
                    problem => {
                        type   => 'checkout',
                        patron => $patron,
                        holds  => \@not_checked_out
                    }
                );
            }
        }

        $curbside_pickup->set(
            {
                arrival_datetime   => $curbside_pickup->arrival_datetime || dt_from_string(),
                delivered_datetime => dt_from_string(),
                delivered_by       => C4::Context->userenv->{number},
            }
        )->store()
          if $curbside_pickup;
    }

    my $curbside_pickups = Koha::CurbsidePickups->search(
        {
            branchcode                => $branchcode,
            scheduled_pickup_datetime => { '>' => \'DATE(NOW())' },
        },
        {
            order_by => { -asc => 'scheduled_pickup_datetime' }
        }
    );

    $template->param(
        branchcode       => $branchcode,
        curbside_pickups => $curbside_pickups,
        pickups_json     => to_json( $curbside_pickups->TO_JSON ),
        tab              => $cgi->param('tab'),
    );
    $self->output_html( $template->output() );
}

sub _notify_new_pickup {
    my ( $self, $pickup ) = @_;

    my $patron = $pickup->patron;

    # Try to get the borrower's email address
    my $to_address = $patron->notice_email_address;

    my $messagingprefs = C4::Members::Messaging::GetMessagingPreferences(
        {
            borrowernumber => $patron->id,
            message_name   => 'Hold_Filled'
        }
    );

    my $library = $pickup->library->unblessed;

    my $admin_email_address = $library->{branchemail}
      || C4::Context->preference('KohaAdminEmailAddress');

    my %letter_params = (
        module     => 'reserves',
        branchcode => $pickup->branchcode,
        lang       => $patron->lang,
        tables     => {
            'branches'  => $library,
            'borrowers' => $patron->unblessed,
        },
        substitute => {
            curbside_pickup => $pickup,
        }
    );

    my $send_notification = sub {
        my ( $mtt, $letter_code, $patron ) = (@_);
        return unless defined $letter_code;
        $letter_params{letter_code}            = $letter_code;
        $letter_params{message_transport_type} = $mtt;
        my $letter = C4::Letters::GetPreparedLetter(%letter_params);
        unless ($letter) {
            warn "Could not find a letter called '$letter_params{'letter_code'}' for $mtt in the 'Holds (reserves)' module";
            return;
        }

        C4::Letters::EnqueueLetter(
            {
                letter                 => $letter,
                borrowernumber         => $patron->id,
                from_address           => $admin_email_address,
                message_transport_type => $mtt,
            }
        );
    };

    while ( my ( $mtt, $letter_code ) =
        each %{ $messagingprefs->{transports} } )
    {
        warn "Curbside Pickup Plugin: borrowernumber " . $patron->id . "requested a notice by email but does not have an email to send to!"
            if $mtt eq 'email' and not $to_address;

        warn "Curbside Pickup Plugin: borrowernumber " . $patron->id . "requested a notice by SMS but does not have an SMS number to send to!"
            if $mtt eq 'sms' and not $patron->smsalertnumber;

        warn "Curbside Pickup Plugin: borrowernumber " . $patron->id . "requested a notice by phone but does not have an phone number to send to!"
            if $mtt eq 'phone' and not $patron->phone;

        next
          if (
            ( $mtt eq 'email' and not $to_address )    # No email address
            or ( $mtt eq 'sms' and not $patron->smsalertnumber ) # No SMS number
            or ( $mtt eq 'phone' and not $patron->phone )    # No phone number
          );

        $send_notification->( $mtt, 'CURBSIDE', $patron );
    }
}

sub opac_js {
    my ($self) = @_;

    return read_file( abs_path( $self->mbf_path('opac.js') ) );
}

sub intranet_js {
    my ($self) = @_;

    return read_file( abs_path( $self->mbf_path('intranet.js') ) );
}

sub configure {
    my ( $self, $args ) = @_;
    my $cgi = $self->{'cgi'};

    my $libraries =
      Koha::Libraries->search( {}, { order_by => ['branchname'] } );

    unless ( $cgi->param('save') ) {

        my $dir = $self->bundle_path.'/i18n';
        opendir(my $dh, $dir) || die "Can't opendir $dir: $!";
        my @files = grep { /^[^.]/ && -f "$dir/$_" } readdir($dh);
        closedir $dh;

        my @tokens;
        foreach my $file (@files) {
            my @splitted = split(/\./, $file, -1);
            my $lang = $splitted[0];
            push @tokens, {key => $lang, text => decode("UTF-8", $self->mbf_read('i18n/'.$file))};
        }

        my $template = $self->get_template( { file => 'configure.tt' } );

        my %policies = map { $_->branchcode => $_ }
          Koha::CurbsidePickupPolicies->search()->as_list;

        $template->param(
            policies  => \%policies,
            libraries => $libraries,
            tokens => \@tokens,
        );

        $self->output_html( $template->output() );
    }
    else {
        foreach my $library ( $libraries->as_list ) {
            my $branchcode = $library->branchcode;

            my $params;

            $params->{branchcode}      = $branchcode;
            $params->{enabled}         = $cgi->param("enable-$branchcode") || 0;
            $params->{pickup_interval} = $cgi->param("interval-$branchcode");
            $params->{patrons_per_interval} =
              $cgi->param("max-per-interval-$branchcode");
            $params->{patron_scheduled_pickup} =
              $cgi->param("patron-scheduled-$branchcode") || 0;

            for my $day (
                qw( sunday monday tuesday wednesday thursday friday saturday ))
            {
                for my $start_end (qw( start end )) {
                    for my $hour_min (qw( hour minute )) {

                        my $value = $cgi->param(
                            "pickup-$start_end-$hour_min-$day-$branchcode");
                        $value = undef if $value eq q{};

                        my $key = $day . '_' . $start_end . '_' . $hour_min;

                        $params->{$key} = $value;
                    }
                }
            }

            my $CurbsidePickupPolicy = Koha::CurbsidePickupPolicies->find(
                { branchcode => $branchcode } );
            $CurbsidePickupPolicy->delete if $CurbsidePickupPolicy;

            Koha::CurbsidePickupPolicy->new($params)->store();
        }
        $self->go_home();
    }
}

sub install() {
    my ( $self, $args ) = @_;

    my $dbh = C4::Context->dbh;

    $dbh->do(q{
CREATE TABLE `curbside_pickup_policy` (
  `id` int(11) NOT NULL auto_increment,
  `branchcode` varchar(10) NOT NULL,
  `enabled` INT(1) NOT NULL DEFAULT 0,
  `pickup_interval` INT(2) NOT NULL DEFAULT 0,
  `patrons_per_interval` INT(2) NOT NULL DEFAULT 0,
  `patron_scheduled_pickup` INT(1) NOT NULL DEFAULT 0,
  `sunday_start_hour` INT(2) NULL DEFAULT NULL,
  `sunday_start_minute` INT(2) NULL DEFAULT NULL,
  `sunday_end_hour` INT(2) NULL DEFAULT NULL,
  `sunday_end_minute` INT(2) NULL DEFAULT NULL,
  `monday_start_hour` INT(2) NULL DEFAULT NULL,
  `monday_start_minute` INT(2) NULL DEFAULT NULL,
  `monday_end_hour` INT(2) NULL DEFAULT NULL,
  `monday_end_minute` INT(2) NULL DEFAULT NULL,
  `tuesday_start_hour` INT(2) NULL DEFAULT NULL,
  `tuesday_start_minute` INT(2) NULL DEFAULT NULL,
  `tuesday_end_hour` INT(2) NULL DEFAULT NULL,
  `tuesday_end_minute` INT(2) NULL DEFAULT NULL,
  `wednesday_start_hour` INT(2) NULL DEFAULT NULL,
  `wednesday_start_minute` INT(2) NULL DEFAULT NULL,
  `wednesday_end_hour` INT(2) NULL DEFAULT NULL,
  `wednesday_end_minute` INT(2) NULL DEFAULT NULL,
  `thursday_start_hour` INT(2) NULL DEFAULT NULL,
  `thursday_start_minute` INT(2) NULL DEFAULT NULL,
  `thursday_end_hour` INT(2) NULL DEFAULT NULL,
  `thursday_end_minute` INT(2) NULL DEFAULT NULL,
  `friday_start_hour` INT(2) NULL DEFAULT NULL,
  `friday_start_minute` INT(2) NULL DEFAULT NULL,
  `friday_end_hour` INT(2) NULL DEFAULT NULL,
  `friday_end_minute` INT(2) NULL DEFAULT NULL,
  `saturday_start_hour` INT(2) NULL DEFAULT NULL,
  `saturday_start_minute` INT(2) NULL DEFAULT NULL,
  `saturday_end_hour` INT(2) NULL DEFAULT NULL,
  `saturday_end_minute` INT(2) NULL DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY (`branchcode`),
  FOREIGN KEY (branchcode) REFERENCES branches(branchcode) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
    });

    $dbh->do(q{
CREATE TABLE `curbside_pickups` (
  `id` int(11) NOT NULL auto_increment,
  `borrowernumber` int(11) NOT NULL,
  `branchcode` varchar(10) NOT NULL,
  `scheduled_pickup_datetime` datetime NOT NULL,
  `staged_datetime` datetime NULL DEFAULT NULL,
  `staged_by` int(11) NULL DEFAULT NULL,
  `arrival_datetime` datetime NULL DEFAULT NULL,
  `delivered_datetime` datetime NULL DEFAULT NULL,
  `delivered_by` int(11) NULL DEFAULT NULL,
  `notes` text NULL DEFAULT NULL,
  PRIMARY KEY (`id`),
  FOREIGN KEY (branchcode) REFERENCES branches(branchcode) ON DELETE CASCADE ON UPDATE CASCADE,
  FOREIGN KEY (borrowernumber) REFERENCES borrowers(borrowernumber) ON DELETE CASCADE ON UPDATE CASCADE,
  FOREIGN KEY (staged_by) REFERENCES borrowers(borrowernumber) ON DELETE SET NULL ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
	});

    $dbh->do(q{
CREATE TABLE `curbside_pickup_issues` (
  `id` int(11) NOT NULL auto_increment,
  `curbside_pickup_id` int(11) NOT NULL,
  `issue_id` int(11) NOT NULL,
  `reserve_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  FOREIGN KEY (curbside_pickup_id) REFERENCES curbside_pickups(id) ON DELETE CASCADE ON UPDATE CASCADE,
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
	});

    $dbh->do(q{
CREATE TABLE `curbside_pickup_issues` (
  `id` int(11) NOT NULL auto_increment,
  `curbside_pickup_id` int(11) NOT NULL,
  `issue_id` int(11) NOT NULL,
  `reserve_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  FOREIGN KEY (curbside_pickup_id) REFERENCES curbside_pickups(id) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
    });

    $dbh->do(q{
        INSERT INTO `letter` VALUES ('reserves','CURBSIDE','','Curbside pickup',0,'You have schedule a curbside pickup for <<branches.branchname>>','[%- USE KohaDates -%]\r\n[%- SET cp = curbside_pickup -%]\r\n\r\nYou have a curbside pickup scheduled for [% cp.scheduled_pickup_datetime | $KohaDates with_hours => 1 %] at [% cp.library.branchname %].\r\n\r\nAny holds waiting for you at the pickup time will be included in this pickup. At this time, that list includes:\r\n[%- FOREACH h IN cp.patron.holds %]\r\n    [%- IF h.branchcode == cp.branchcode && h.found == \'W\' %]\r\n* [% h.biblio.title %], [% h.biblio.author %] ([% h.item.barcode %])\r\n    [%- END %]\r\n[%- END %]\r\n\r\nOnce you have arrived, please call your library or log into your account and click the \"Alert staff of your arrival\" button to let them know you are there.','email','default');
    });

    return 1;
}

sub upgrade {
    my ( $self, $args ) = @_;

    my $dbh = C4::Context->dbh;

    unless ( TableExists('curbside_pickup_issues') ) {
        $dbh->do(q{
CREATE TABLE `curbside_pickup_issues` (
  `id` int(11) NOT NULL auto_increment,
  `curbside_pickup_id` int(11) NOT NULL,
  `issue_id` int(11) NOT NULL,
  `reserve_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  FOREIGN KEY (curbside_pickup_id) REFERENCES curbside_pickups(id) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
    	});
    }

    return 1;
}

sub uninstall() {
    my ( $self, $args ) = @_;

}

sub api_routes {
    my ( $self, $args ) = @_;

    my $spec_str = $self->mbf_read('openapi.json');
    my $spec     = decode_json($spec_str);

    return $spec;
}

sub api_namespace {
    my ($self) = @_;

    return 'curbsidepickup';
}

1;
